pkg/tfbridge/detailed_diff_test.go

```diff
@@ -0,0 +1,1631 @@
+			want: &pulumirpc.PropertyDiff{Kind: pulumirpc.PropertyDiff_UPDATE},
+		},
+		{
+			name: "changed from nil",
+			old:  resource.NewNullProperty(),
+			new:  resource.NewStringProperty("new"),
+			want: &pulumirpc.PropertyDiff{Kind: pulumirpc.PropertyDiff_ADD},
+		},
+		{
+			name: "changed to nil",
+			old:  resource.NewStringProperty("old"),
+			new:  resource.NewNullProperty(),
+			want: &pulumirpc.PropertyDiff{Kind: pulumirpc.PropertyDiff_DELETE},
+		},
+		{
+			name: "tf force new unchanged",
+			old:  resource.NewStringProperty("old"),
+			new:  resource.NewStringProperty("old"),
+			etf:  shimschema.Schema{ForceNew: true},
+			want: nil,
+		},
+		{
+			name: "tf force new changed non-nil",
+			old:  resource.NewStringProperty("old"),
+			new:  resource.NewStringProperty("new"),
+			etf:  shimschema.Schema{ForceNew: true},
+			want: &pulumirpc.PropertyDiff{Kind: pulumirpc.PropertyDiff_UPDATE_REPLACE},
+		},
+		{
+			name: "tf force new changed from nil",
+			old:  resource.NewNullProperty(),
+			new:  resource.NewStringProperty("new"),
+			etf:  shimschema.Schema{ForceNew: true},
+			want: &pulumirpc.PropertyDiff{Kind: pulumirpc.PropertyDiff_ADD_REPLACE},
+		},
+		{
+			name: "tf force new changed to nil",
+			old:  resource.NewStringProperty("old"),
+			new:  resource.NewNullProperty(),
+			etf:  shimschema.Schema{ForceNew: true},
+			want: &pulumirpc.PropertyDiff{Kind: pulumirpc.PropertyDiff_DELETE_REPLACE},
+		},
+		{
+			name: "ps force new unchanged",
+			old:  resource.NewStringProperty("old"),
+			new:  resource.NewStringProperty("old"),
+			eps:  &SchemaInfo{ForceNew: True()},
+			want: nil,
+		},
+		{
+			name: "ps force new changed non-nil",
+			old:  resource.NewStringProperty("old"),
+			new:  resource.NewStringProperty("new"),
+			eps:  &SchemaInfo{ForceNew: True()},
+			want: &pulumirpc.PropertyDiff{Kind: pulumirpc.PropertyDiff_UPDATE_REPLACE},
+		},
+		{
+			name: "ps force new changed from nil",
+			old:  resource.NewNullProperty(),
+			new:  resource.NewStringProperty("new"),
+			eps:  &SchemaInfo{ForceNew: True()},
+			want: &pulumirpc.PropertyDiff{Kind: pulumirpc.PropertyDiff_ADD_REPLACE},
+		},
+		{
+			name: "ps force new changed to nil",
+			old:  resource.NewStringProperty("old"),
+			new:  resource.NewNullProperty(),
+			eps:  &SchemaInfo{ForceNew: True()},
+			want: &pulumirpc.PropertyDiff{Kind: pulumirpc.PropertyDiff_DELETE_REPLACE},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			got := detailedDiffer{
+				tfs: shimschema.SchemaMap{"foo": tt.etf.Shim()},
+				ps:  map[string]*SchemaInfo{"foo": tt.eps},
+			}.makeTopPropDiff(tt.old, tt.new, newPropertyPath("foo"))
+			if got == nil && tt.want == nil {
+				return
+			}
+			if got == nil || tt.want == nil {
+				t.Errorf("makeTopPropDiff() = %v, want %v", got, tt.want)
+				return
+			}
+			if got.Kind != tt.want.Kind {
+				t.Errorf("makeTopPropDiff() = %v, want %v", got.String(), tt.want.String())
+			}
```

Comment:
Why can't we just `assert.Equal(t, tt.want, got)`.

Nit: testify calls these "expected" and "actual". I like that better then "want" and "got".

Reply:
```
expected: map[tfbridge.detailedDiffKey]*pulumirpc.PropertyDiff(map[tfbridge.detailedDiffKey]*pulumirpc.PropertyDiff{"foo":(*pulumirpc.PropertyDiff)(0x140009a6e40)})
                                actual  : *pulumirpc.PropertyDiff(&pulumirpc.PropertyDiff{state:impl.MessageState{NoUnkeyedLiterals:pragma.NoUnkeyedLiterals{}, DoNotCompare:pragma.DoNotCompare{}, DoNotCopy:pragma.DoNotCopy{}, atomicMessageInfo:(*impl.MessageInfo)(nil)}, sizeCache:0, unknownFields:[]uint8(nil), Kind:1, InputDiff:false})
```

The maps are not comparable but I've simplified it a bit here, thanks.

Reply:
What do you mean by "not comparable"? They are not `comparable`, but `assert` handles that with reflection. You can `assert.Equal` on 2 maps.

If you mean comparable outside of Go's definition, can you explain?

Reply:
I had misread the error message here, assuming the assert library does not work for these types. Thanks for following up.